cfep-13.md

```diff
@@ -0,0 +1,70 @@
+
+<table>
+<tr><td> Title </td><td> Securing `conda-forge` Uploads to `anaconda.org` </td>
+<tr><td> Status </td><td> Proposed </td></tr>
+<tr><td> Author(s) </td><td> Matt, Marius, Anthony, Isuru</td></tr>
+<tr><td> Created </td><td> Feb 27, 2020</td></tr>
+<tr><td> Updated </td><td> Feb 27, 2020</td></tr>
+<tr><td> Discussion </td><td> NA </td></tr>
+<tr><td> Implementation </td><td> NA </td></tr>
+</table>
+
+## Abstract
+
+Currently the `anaconda.org` tokens on each feedstock allow any `conda-forge`
+feedstock maintainer to upload any output to any `anaconda.org` package.
+This could allow someone to push a compromised and/or malicious build
+of say the `python` package. This situation is a security risk that we
+should mitigate.
+
+## Specification
+
+We will secure the uploads of outputs to `anaconda.org` via the following process.
+
+1. Each feedstock is provisioned with a secret key that is only known to the
+   maintainers of that feedstock and the core team. These keys will be kept in
+   a private repo with only core members having write access. This key will be pushed to
+   the CI providers for that repo as well.
+2. Feedstock creation will move to a separate process in order to secure the keys
```

Comment:
What does this mean? Why can't we have the same process as now?

Reply:
Well @mariusvniekerk seemed to suggest on gitter that this would be a good change, but I agree the exact technical issue here is not clear to me. 

We do need to generate and register new keys to the private repo in a secure way. IDK if having this done with code on staged recipes is a great idea. 

However, we don't have to redo staged recipes to address this. We could have the web service watch repo creation events in the conda-forge org and automatically register a key when it detects a new feedstock.

Reply:
We also need to add the keys to feedstock CI services in a secure way. Again it is not 100% clear to me how we do this. Maybe the webservice has a register command we call.

Reply:
Why can't we have the feedstock creation script in `staged-recipes` do this?

Reply:
The actual constraint is that we have to avoid leaking the secret key we generate to the outside world. 

The question of who can see those keys is separate at some level. Having them be generated in staged recipes would allow core+any staged recipe maintainer be able to see the keys as they are made in principle.

If we are ok with that, then staged-recipes is fine. 

After thinking about it for a while and remembering that I was not always on core, I think I'd still prefer to not have even our encrypted tokens for any service available outside of core. However, I agree this is a different issue that is orthogonal to this CFEP.

Reply:
I agree that this is orthogonal to this CFEP. (If we go this route, we'd need to remove the feedstock creation scripts from staged-recipes as they contain github tokens, CI tokens, etc.)